### PR TITLE
Test Driven Development

### PR DESCRIPTION
**Pull request by jagjennings**

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -11,7 +11,7 @@
     <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.0/js/bootstrap.min.js"></script>
     <script src="script.js"></script>
   </head>
-  <body onload="getComments()" class="container-fluid pt-3 pb-3">
+  <body onload="getComments(); getLogin();" class="container-fluid pt-3 pb-3">
```
Comment (qihuang2) [inline portfolio/src/main/webapp/index.html]: can we move these changes into a separate PR. this doesn't seem to be related to the TDD stuff

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,83 @@
...
+
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if (!mandatoryAttendees.isEmpty()) {
+      return timeRanges;
+    } else {
+      return new ArrayList<TimeRange>();
+    }
+  }
+
+  public Collection<TimeRange> findTimeRanges(ArrayList<String> attendees, Collection<Event> events, MeetingRequest request) {
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: let's rename this to something more descriptive, like findAvailableTimeRanges

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,83 @@
...
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: add a javadoc for what this method does: https://google.github.io/styleguide/javaguide.html#s7-javadoc

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: looks like MeetingRequest is only used for getting duration. we can just pass in duration instead of the  MeetingRequest object

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,83 @@
...
+        if (event.getAttendees().contains(attendee)) {
+          badTimeRanges.add(event.getWhen());
+        }
+      }
+    }
+
+    int availableStartTime = TimeRange.START_OF_DAY;
+
+    // Store all time ranges between unavailable time ranges that fit the duration of the meeting
+    for (TimeRange timeRange: badTimeRanges) {
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: are we assuming badTimeRanges is sorted? not sure if this will work if ranges aren't sorted

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,83 @@
...
+        availableStartTime = timeRange.end();
+      }
+    }
+
+    // Include time from the end of the last unavailable time range to the end of the day
+    if (TimeRange.END_OF_DAY - availableStartTime >= duration) {
+      timeRanges.add(TimeRange.fromStartEnd(availableStartTime, TimeRange.END_OF_DAY, true));
+    }
+
+    // Check result for overlap with unavailable times and remove overlapping times
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: in what situation will there be overlaps?

**Review by qihuang2 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (jagjennings) [reply portfolio/src/main/webapp/index.html]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: done. sorted  badTimeRanges by Start Time using Comparator.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: now that the badTimeRanges are sorted, there is no longer a need to check for overlap. removed.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (qihuang2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: is this comment correct? I thought the method returns TimeRanges that events can be scheduled for

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,86 @@
...
+
+    // If no time ranges are available, exclude optional attendees and try again
+    if (!timeRanges.isEmpty()) {
+      return timeRanges;
+    } else {
+      timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(mandatoryAttendees, events, duration));
+    }
+
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if (!mandatoryAttendees.isEmpty()) {
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: should we do this mandatoryAttendees.isEmpty() check before calling timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(mandatoryAttendees, events, duration));? 

right now it's making an unnecessary call to findAvailableTimeRanges if mandatoryAttendees are empty. also another question, should we return Arrays.asList(TimeRange.WHOLE_DAY) if there are optional attendees but no mandatory attendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,86 @@
...
+    // Make a list of all unavailable time ranges for all attendees
+    for (Event event: events) {
+      for (String attendee: attendees) {
+        if (event.getAttendees().contains(attendee)) {
+          badTimeRanges.add(event.getWhen());
+        }
+      }
+    }
+
+    if (!badTimeRanges.isEmpty()) {
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: we can remove this badTimeRanges.isEmpty(). Collections.sort no-ops if list to sort is empty

**Review by qihuang2 — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: added the check for mandatoryAttendees before making the call.

I agreed with your logic on how the code should handle requests with optional attendees but no mandatory ones, but the walkthrough instructed us to return an **empty list** in these situations. Without this portion of the code, it would follow the same logic you described.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (qihuang2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: ahh i see, LG then!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,83 @@
...
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
+
+  /**
+  * Determines which attendees to include and returns time ranges that accomadate those attendees
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: can you add some more comments about the edge cases? like if there are no mandatory attendees but are optional attendees, what happens? should also mention that optional attendees are first considered when finding available times, but they are removed from consideration if no available times are found with them considered

javadocs are useful for people that might come along later, and have no idea what this method is supposed to do. they can read the javadoc and know what the method promises to do without having to read the source code. stating what happens during edge cases is extremely helpful

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,83 @@
...
+    ArrayList<TimeRange> timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(allAttendees, events, duration));
+
+    // If no time ranges are available, exclude optional attendees and try again
+    if (!timeRanges.isEmpty()) {
+      return timeRanges;
+    }
+
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if (!mandatoryAttendees.isEmpty()) {
+      timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(mandatoryAttendees, events, duration));
```
Comment (qihuang2) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: you can just do 
return new ArrayList<TimeRange>(findAvailableTimeRanges(mandatoryAttendees, events, duration));

**Review by qihuang2 — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (jagjennings) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.
